Create a setting for user language preference

0dysseas opened the pull request:
<!-- What issue does this PR close? -->
Closes #2815 

<!-- What does this PR achieve? [feature|hotfix|fix|refactor] -->
This implements a new website language selection option, positioned at the footer. The available languages for the user to select are the ones mentioned in the #2815 ticket (i.e. Čeština, Deutsch, English, Español, Français and Telugu), with the default option being English.

### Technical
<!-- What should be noted about the implementation? -->
This feature is implemented using cookies in order for any type of visitor, whether logged-in or not, to be able to use it. 

More specifically, a new cookie titled `i18n_code` is now being sent from the server. As no explicit expiration date has been set, the cookie is valid for a user session (although this can be changed if deemed necessary). Then, the user upon clicking on any of the new available languages (found in the navigation footer) will change the value of the cookie and perform a new request to the server which will then render the website in the user selected language.

For the scope of this ticket the `.pot` file was updated with the new text used for the language settings. However, no new language files are committed with this PR since I cannot vouch for the validity of the translations.

Finally, the tests have been updated accordingly in order to accommodate for the new functionality.

### Testing
<!-- Steps for reviewer to reproduce/verify what this PR does/fixes. -->
Just navigate to the bottom of the website and click one of the available languages under the "Change Website Language" section.

### Screenshot
<!-- If this PR touches UI, please post evidence (screenshots) of it behaving correctly. -->

New feature:
![with_website_language](https://user-images.githubusercontent.com/31179964/102691132-ec864880-4212-11eb-8710-2923f11da3bf.png)

Spanish language is selected:
![spanish_selected](https://user-images.githubusercontent.com/31179964/102691138-02940900-4213-11eb-9bec-ea206e73facb.png)

and

![spanish_selected_2](https://user-images.githubusercontent.com/31179964/102691141-058ef980-4213-11eb-9729-9dee9a636933.png)

### Stakeholders
<!-- @ tag stakeholders of this bug -->

cclauss commented on openlibrary/core/helpers.py:
```diff
@@ -122,26 +124,32 @@ def days_since(then, now=None):
     """Internationalized version of web.datestr."""
-    lang = lang or web.ctx.get('lang') or "en"
+    # Get the locale from the cookie or default to 'en'
+    locale = web.cookies(i18n_code='en')
```
Should we be tapping into https://docs.python.org/3/library/locale.html ?

tfmorris replied:
No. You don't want the server's locale.

0dysseas replied:
In any case, this is a good point. I'll go ahead and rename the `locale` variable to avoid any future name-clashing (in case someone imports the `locale` module)

cclauss commented on openlibrary/core/helpers.py:
```diff
@@ -109,9 +109,11 @@ def safesort(iterable, key=None, reverse=False):
     care to make that work.
     """
     key = key or (lambda x: x)
+
```
These code reformatting changes distract the reviewer when trying to understand the changes.   I would be happy to see the whole codebase formatted with `psf/black --skip-string-normalization` in a ___separate___ PR but let's not do it across multiple PRs.  So, please only make formatting changes in this PR if the on the lines that this PR is modifying.

0dysseas replied:
Ok, sure, will do.

tfmorris commented on openlibrary/plugins/openlibrary/home.py:
```diff
@@ -51,22 +51,32 @@ def get_homepage():
+        if locale_cookie is not None:
+            locale = locale_cookie
+        else:
+            web.setcookie('i18n_code', 'en', secure=False)
```
If the user hasn't selected a language yet, the default should be the highest priority language from their browser's `Accept-Language` header which is supported by OpenLibrary. https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Accept-Language

There's no reason to force them to start with English when they've already expressed their language preferences.

0dysseas replied:
Ok, I'll look into it and provide a fix for this. 

If the highest priority language is not part of the supported set of languages (EN, DE, CS, ES, FR, TE) the setting should default to English, right?

0dysseas replied:
Hi @tfmorris. I've implemented an updated fix where the language is selected from the corresponding header. If no OL supporting language is found or if no `Accept-Language` header is sent then, English is used. Is this OK? Or is something more that you want me to do before submitting the updated PR?

tfmorris replied:
English is fine as a fallback if OpenLibrary doesn't support any of their preferred languages.

tfmorris commented on openlibrary/plugins/openlibrary/home.py:
```diff
@@ -51,22 +51,32 @@ def get_homepage():
-        cached_homepage = get_cached_homepage()
+        # Get the locale from the cookie or -if not found- set it to the default ('en')
+        locale_cookie = web.cookies().get('i18n_code')
+        if locale_cookie is not None:
```
```suggestion
        if locale_cookie:
```
I think the `is not None` is redundant, but @cclauss would know for sure.

0dysseas replied:
For readability reasons  prefer the explicit check to `is not None`, but I have no strong opinion on this. Let me know what works best for you.

cclauss replied:
Unfortunately, we still need support legacy Python so we can not use Python type hints.  How would we want the code to behave if `locale_cookie` was any other _falsey_ value like [], {}, set(), tuple(), 0, 0.0, False, etc.?  My sense is that all _falsey_ values are just as dangerous as None.

0dysseas replied:
Correct- all _falsey_ values should fall under that `if` clause. However, according to the official web.py (documentation)[], `web.cookies(https://webpy.org/cookbook/cookies).get(cookieName)` should return only `None`. Thus, the explicit check against that value.

0dysseas replied:
Hi, @cclauss. Is it OK if I leave this as is or should I go ahead and remove the `is not None` check?

0dysseas replied:
As this change is not a show-stopper and in an effort to merge the updated PR as soon as possible, I'm leaving this as is. Please don't hesitate to comment on the new PR, should you think that a change here is mandatory.

tfmorris commented on openlibrary/templates/lib/nav_foot.html:
```diff
@@ -47,6 +47,17 @@ <h2>$:_('Help')</h2>
+          <li id="en"><a href="/" title="$_('English')">English (en)</a></li>
+          <li id="es"><a href="/" title="$_('Spanish')">Español (es)</a></li>
+          <li id="fr"><a href="/" title="$_('French')">Français (fr)</a></li>
+          <li id="te"><a href="/" title="$_('Telugu')">Telugu (te)</a></li>
```
```suggestion
          <li id="te"><a href="/" title="$_('Telugu')">తెలుగు (te)</a></li>
```

0dysseas replied:
Ok, will change this

tfmorris commented on openlibrary/templates/lib/nav_foot.html:
```diff
@@ -47,6 +47,17 @@ <h2>$:_('Help')</h2>
+      <div id="footer-locale-menu">
+        <h2>$:_('Change Website Language')</h2> 
+        <ul id="locale-options">
+          <li id="cs"><a href="/" title="$_('Czech')">Čeština (cs)</a> </li>
+          <li id="de"><a href="/" title="$_('German')">Deutsch (de)</a> </li>
+          <li id="en"><a href="/" title="$_('English')">English (en)</a></li>
+          <li id="es"><a href="/" title="$_('Spanish')">Español (es)</a></li>
+          <li id="fr"><a href="/" title="$_('French')">Français (fr)</a></li>
+          <li id="te"><a href="/" title="$_('Telugu')">Telugu (te)</a></li>
```
I don't understand why the title is being translated, but the actual text isn't. Is that an oversight? Actually, since they should all be autoglottonyms, they will be the same in every language, so don't need to be included in the translation files.

0dysseas replied:
[My understanding](https://github.com/internetarchive/openlibrary/tree/master/openlibrary/i18n#updating-an-existing-language) was that any new string provided should be included in the default `messages.pot` file but it is acceptable to not commit the translated text if one's unsure about it.

So I added the title as well as the Czech and Telugu strings (the rest of the languages were already in the `messages.pot` file) but didn't provide their translated counterparts as I were unsure about their validity.

Should I completely remove them from the `messages.pot` file?

tfmorris replied:
Yes, I think you can use the same string for both the text and the title e.g.
```
          <li id="cs"><a href="/" title="Čeština">Čeština (cs)</a> </li>
```
It doesn't need to be translated, because users will be expecting to find it in their native language ie not "Czech," which they might not recognize if they don't speak English.

tfmorris commented on openlibrary/plugins/openlibrary/home.py:
```diff
@@ -86,20 +96,39 @@ def get_ia_carousel_books(query=None, subject=None, work_id=None, sorts=None,
-        'art', 'science_fiction', 'fantasy', 'biographies', 'recipes',
-        'romance', 'textbooks', 'children', 'history', 'medicine', 'religion',
-        'mystery_and_detective_stories', 'plays', 'music', 'science'
+        'art',
+        'science_fiction',
+        'fantasy',
+        'biographies',
+        'recipes',
+        'romance',
+        'textbooks',
+        'children',
+        'history',
+        'medicine',
+        'religion',
+        'mystery_and_detective_stories',
+        'plays',
+        'music',
+        'science',
```
No unrelated formatting noise please.

0dysseas replied:
Ok, sure, will revert this back to the original format.

tfmorris commented on openlibrary/i18n/messages.pot:
```diff
@@ -2556,6 +2556,18 @@ msgstr ""
+msgid "Change Website Language"
+msgstr ""
+
+#: nav_foot.html:53
+msgid "Czech"
+msgstr ""
+
+#: nav_foot.html:58
+msgid "Telugu"
+msgstr ""
+
```
As I mentioned in another comment, these can probably be skipped (at least for use in the footer).

0dysseas replied:
I thought that adding new text in the `messages.pot` file was standard procedure for all new [user visible strings](https://github.com/internetarchive/openlibrary/wiki/Frontend-Guide#internationalization-i18n---for-translators). Should I go ahead and remove them?

tfmorris replied:
Yes, please remove them. They don't need to be translated, per my comment above.

tfmorris commented on openlibrary/tests/core/test_i18n.py:
```diff
@@ -13,24 +15,40 @@ def ungettext(self, message1, message2, n):
     def test_ungettext(self, monkeypatch):
         self.setup_monkeypatch(monkeypatch)
 
+        web.setcookie('i18n_code', 'es')
```
I don't see anything that tests if this causes the behavior to change.

0dysseas replied:
Sorry, I don't get the point. I added the code above in order for the test of the `ungettext` function to successfully pass. I believe, that the actual behavior of the function remains the same. 

Prior to my commit, the language was supposed to be accessed via the URL (`load_translations(web.ctx.get('lang', 'en'))`) whereas with my changes the language is retrieved from the cookies. Both, actions are done using inherent `web.py` functionality.

tfmorris replied:
OK, got it.

tfmorris commented on openlibrary/tests/core/test_i18n.py:
```diff
@@ -13,24 +15,40 @@ def ungettext(self, message1, message2, n):
 
         monkeypatch.setattr(i18n, "load_translations", self.d)
         monkeypatch.setattr(web, "ctx", ctx)
+        self._load_fake_context()
```
Some tests of the HTTP Accept-Language header are needed, including ones that request languages which aren't supported or where the first supported language is 2nd or 3rd on the list.

0dysseas replied:
Ok. Yes, I will try to provide tests for the new solution.

0dysseas replied:
New tests for the `Accept-Language` header will be provided (including the tests you've mentioned in your comment).

tfmorris reviewed (changes requested):
Thanks for working on this! It's been a problem for years.

I'll second @cclauss request to cut down on the noise in the diff by removing unrelated changes.

Rather than defaulting to English, it should default to the language in the `Accept-Language` header from the user's browser, **but** I think this may already be handled by Infogami, so I'd check that out before making any code changes.

tfmorris commented:
It looks like my commit privileges got revoked over the holidays, so whatever @cclauss says goes.

![Screen Shot 2021-01-09 at 12 44 17 PM](https://user-images.githubusercontent.com/82178/104104811-65a81700-5278-11eb-9659-1092001e9112.png)

0dysseas commented:
Cool. Will implement @cclauss  suggestion and push again.

cdrini commented on openlibrary/plugins/openlibrary/home.py:
```diff
@@ -51,17 +52,20 @@ def get_homepage():
 
-def get_cached_homepage():
+
+def get_cached_homepage(lang='en'):
```
This no longer needs to be optional.

```suggestion
def get_cached_homepage(lang):
```

0dysseas replied:
Done

cdrini commented on openlibrary/templates/lib/nav_foot.html:
```diff
@@ -47,6 +47,17 @@ <h2>$:_('Help')</h2>
+      <div id="footer-locale-menu">
+        <h2>$:_('Change Website Language')</h2> 
+        <ul id="locale-options">
+          <li id="cs"><a href="/" title="Čeština">Čeština (cs)</a> </li>
+          <li id="de"><a href="/" title="Deutsch">Deutsch (de)</a> </li>
+          <li id="en"><a href="/" title="English">English (en)</a></li>
+          <li id="es"><a href="/" title="Español">Español (es)</a></li>
+          <li id="fr"><a href="/" title="Français">Français (fr)</a></li>
+          <li id="te"><a href="/" title="తెలుగు">తెలుగు (te)</a></li>
```
These `id` are too short; they're likely to conflict. We can use a data param:

```suggestion
          <li><a href="#" data-lang-id="cs">Čeština (cs)</a></li>
          <li><a href="#" data-lang-id="de">Deutsch (de)</a></li>
          <li><a href="#" data-lang-id="en">English (en)</a></li>
          <li><a href="#" data-lang-id="es">Español (es)</a></li>
          <li><a href="#" data-lang-id="fr">Français (fr)</a></li>
          <li><a href="#" data-lang-id="te">తెలుగు (te)</a></li>
```

Also: It's best practice to not include a `title` if it's the same as the text content, since screenreaders might read it twice.

Also instead of going to the homepage, we likely want these to remain on the current page. So I set the `href` to `#`, and moved the `data-lang-id` to the `<a>` element so we can bind a click event handler to that, and then use `event.preventDefault()` and just trigger a `window.reload()` instead.

0dysseas replied:
Done

cdrini commented on openlibrary/plugins/openlibrary/js/ol.js:
```diff
@@ -147,3 +155,13 @@ export function initPreviewButton() {
+
+export function initWebsiteTranslationOptions() {
+    $(document).ready(function () {
+        $('#locale-options li').on('click', function () {
+            var locale = $(this).attr('id');
+            setValueInCookie('i18n_code', locale);
```
```suggestion
        $('#locale-options li a').on('click', function (event) {
            event.preventDefault();
            const locale = $(this).data('lang-id');
            setValueInCookie('i18n_code', locale);
            location.reload();
```

0dysseas replied:
Done

cdrini commented on openlibrary/i18n/__init__.py:
```diff
@@ -101,6 +103,34 @@ def load_translations(lang):
         return Translations(open(mo_path, 'rb'))
 
+
+def get_ol_locale():
```
To make this a little easier to test (i.e. avoid having to mock `web.ctx`), I often do something like this:

```suggestion
def get_ol_locale(locale_cookie=web.cookies().get('i18n_code'), browser_accepted_langs=web.ctx.env.get('HTTP_ACCEPT_LANGUAGE')):
```

That way the actual code can easily use this function, but in the tests, if I want to test specific flows, I can override them easily.

0dysseas replied:
Done. Using the `i18n_loadhook` now.

cdrini commented on openlibrary/i18n/__init__.py:
```diff
@@ -101,6 +103,34 @@ def load_translations(lang):
+                    break
+        else:
+            ol_locale = 'en'
+        web.setcookie('i18n_code', ol_locale, secure=False)
```
I don't think we want to set this cookie unless the user explicitly clicks the button. If the user changes their browser settings, and they haven't explicitly set the language, we DO want OL to use the right language. (This also makes testing easier :P )

```suggestion
```

0dysseas replied:
Done

cdrini commented on openlibrary/i18n/__init__.py:
```diff
@@ -12,6 +12,8 @@
 
 root = os.path.dirname(__file__)
 
+OL_SUPPORTED_LANGAGES = ['en', 'cs', 'de', 'es', 'fr', 'te']
```
This is going to fall out of sync ; in fact it already is! It's missing `hr`. Let's update and memoize the `get_locales` function to checks the children directories of this directory that contain `messages.po`. Memoize similar to `load_translations`. Might be able to use `glob`? Not sure how to make it use this file's directory (`root` above).

```py
>>> glob.glob('*/*messages.po')
['cs\\messages.po', 'de\\messages.po', 'es\\messages.po', 'fr\\messages.po', 'hr\\messages.po', 'te\\messages.po']
```

cdrini commented on openlibrary/i18n/__init__.py:
```diff
@@ -101,6 +103,34 @@ def load_translations(lang):
+    if locale_cookie:
+        ol_locale = locale_cookie
+    else:
+        browser_accepted_langs = web.ctx.env.get('HTTP_ACCEPT_LANGUAGE')
+        if browser_accepted_langs:
+            # Parse the accepted language header string into a list
+            languages = browser_accepted_langs.split(',')
+            # Remove the quality-value weights and keep only the locales
+            accepted_langs = [lang.partition(';')[0] for lang in languages]
+            # The default locale is 'en' unless we support a browser-suggested language
+            ol_locale = 'en'
+            for lang in accepted_langs:
+                if lang in OL_SUPPORTED_LANGAGES:
+                    ol_locale = lang
+                    break
```
I believe this is already done be `web.ctx.get('lang', 'en')`; it should be checking `HTTP_ACCEPT_LANGUAGE` as well as the url parameter e.g. `?lang=fr` we use for testing.


Oh my... Now that I look at it looks like this is _all_ already done! The cookie is called `HTTP_LANG`. So all we have to do is set it, and everything should work? https://github.com/internetarchive/infogami/blob/03b7f1c38bae80c4930443f46a32b769cdd5fe57/infogami/utils/i18n.py#L132 Ahhh, sorry for the wild goose chaise :(

0dysseas replied:
Done. Using the `i18n_loadhook` now.

cdrini reviewed (changes requested):
This looks awesome, @0dysseas ! Thank you so much for working on this :) Unfortunately I have some annoying news 😓 It looks like a large chunk of this has already been implemented, but we didn't know about it. The cookie `HTTP_LANG` seems like it might do the same thing as the cookie you've introduced, but there's currently no way from the UI to set it.

This section here is what's responsible for setting `web.ctx.lang`, and it looks like it takes into account a cookie: https://github.com/internetarchive/infogami/blob/03b7f1c38bae80c4930443f46a32b769cdd5fe57/infogami/utils/i18n.py#L118-L149

Would you be able to open a new PR and see if using that cookie, with the HTML links + JS you've added, does the trick? Most of my python comments here no longer really apply, since I found this halfway through my code review :( But we wouldn't need any of the new tests or other logic.

Sorry again for the run-around :( That's one of the issues on working on such large and old codebase as open library!

0dysseas commented:
Thanks a lot for your comments @cdrini! Don't worry about the run-around, I totally understand. 

So, I tried your suggestion (i.e. combining my front-end code with the already-existing `i18n_loadhook` function) but, unfortunately, didn't work. Digging into the problem, it seems the `HTTP_LANG` cookie is never set on the server side thus it's inaccessible from the front-end code. I believe a solution to this would require some back-end changes similar to the ones I've introduced (i.e. the `get_ol_locale` function).

I tried refactoring, then, my `get_ol_locale` function to use the `i18n_loadhook`. However, by using `i18n_loadhook()` there's no way to know if/when the language that's returned was set in the `Accept-Language` header or in the cookie. I think, this information is required to properly decide on the supported OL language (e.g. check what's the header language we support -which might not always be the first option- etc).

Is there something that I might've missed? 

Since it seems that additional back-end logic is required to properly set the cookie and also the `i18n_loadhook` function cannot be used as-is, my suggestion is to continue with the current PR by implementing your comments and CRs. But, of course, you have a better overview on this than me so let me know what you think and how I should proceed.

mekarpeles commented:
@0dysseas this is an amazing step in the right direction, thank you for your hard work, completely agree w/ @cdrini + @tfmorris 

thank you @cdrini for your archaeology work + review.

Some design feedback (for when we get there). In terms of the design (footer), can we instead make this a single language dropdown on the topmost Internet Archive black bar? This would hopefully make it more obvious to patrons that the option exists and take less real estate.

![dev openlibrary org_books_OL1005276M_Emotional_development_and_emotional_intelligence (1)](https://user-images.githubusercontent.com/978325/105573292-53ff4e80-5d11-11eb-8734-b55797802295.png)

0dysseas commented:
Thank you @mekarpeles.

I will try to implement your suggestion while waiting for the feedback from @cdrini.

cdrini commented:
Hi @0dysseas ! My initial thoughts are:

1) "it seems the HTTP_LANG cookie is never set on the server side thus it's inaccessible from the front-end code."
  - That should be ok. The front-end shouldn't really ever have to "read" the cookie; it will only ever need to write to it. So it should be ok if the cookie is only _set_ by the client-side code.
2) "by using i18n_loadhook() there's no way to know if/when the language that's returned was set as part of the Accept-Language header or selected by the user in the UI."
  - I think that's ok. 
3) Agreed, we lose the ability to check our supported languages with the old code :/ But I think that's ok. We can always expand that at some point in the future.

Let me know if there are any other questions! If that works, then it should be a better solution, but if there are more issues, we can look at accepting the PR as-is to move things along!

0dysseas commented:
Hi @cdrini,

Thank you for your comments. So, I engineered a solution taking into consideration your points above, using only the `i18n_loadhook()` function and not setting the cookie on server side. Unfortunately, there were some problems with this approach, pertaining to the implementation of the `i18n_loadhook()` function. More specifically:

1) The section of the function responsible for getting the cookie values (e.g. `if "HTTP_LANG" in web.ctx.get("HTTP_COOKIE", "") ) does not work as expected. This piece of code does not retrieve the cookie value (tested on FF and Chrome) and as a result the associated `cookies` variable remains empty. To properly access the cookie value one can use the `web.cookies().get('HTTP_LANG')` function (already used as such in my `get_ol_locale()` function) as described in the [web.py docs](https://webpy.org/cookbook/cookies) . Please let me know if I'm missing something here.

2) If the user has not set the HTTP language header (i.e. no language preferences are set at the browser) and the cookie is to be set only on client-side but the user never selects an OL language, the `i18n_loadhook()` function will return the `lang` query string (if set) or `" "`'.  What should be done in this case? What's the language that should be used across OL? 
The section responsible for this is: `parse_query_string() or parse_lang_cookie() or parse_lang_header() or ' '`. 

I believe that in order to account for the above two cases, the `i18n_loadhook` function will have to be modified to some extent (at least for point #1)  but I'm not sure if this is how you want to move forward with this PR. 

I'm sorry for the back and forth but I'd like to make sure that I code something that will be useful to the community and will not introduce some unwanted functionality. Please let me know what you think the next steps should be.

0dysseas commented:
Hi @mekarpeles,

While waiting @cdrini  's feedback on my last comment, I tried implementing your suggested UI change. However, the topmost Internet Archive black bar is not displayed in my local dev version setup. 

Do I have to explicitly configure something to have this rendered? I looked into the OL documentation but didn't find anything related to this.

cdrini commented:
1. It looks like `i18n_loadhook` does use web.cookies:

https://github.com/internetarchive/infogami/blob/03b7f1c38bae80c4930443f46a32b769cdd5fe57/infogami/utils/i18n.py#L132-L137

The comment suggests it's using `"HTTP_LANG" in web.ctx.get("HTTP_COOKIE", "")` for performance reasons: `# Quick check to avoid making cookies call`

Ahhh, I think I see the issue. I tried setting the cookie via JS (`document.cookie = "HTTP_LANG=fr";`) and then going to a `debug=true` page with an error on it (which prints the environment), and saw this:


Variable | Value
-- | --
HTTP_COOKIE | ('donation-identifier=*****=; pd=; '  'session=/people/ScarTissue******; '  'HTTP_LANG=fr')


~It looks like this is a tuple, not a string, so `in` is not going to work? So then in theory what we want is `i18n_loadhook` to say `if "HTTP_LANG" in str(web.ctx.get("HTTP_COOKIE", "")):`....?~

**Oh goodness, it needs to be `web.ctx.env.get` -_-.** That's why it's not working. I'll open a quick PR to fix that.

For 2, that's effectively the same as setting `en` I believe. All the i18n functions will fallback to `en` if an invalid language (i.e. `' '`) is provided. Note it's _very_ unlikely that `HTTP_ACCEPT_LANGUAGE` will be unset; most browsers set that.

cdrini commented:
That should be fixed now. You'll need to do a rebase and `docker-compose up --pull` once the docker image is done (likely ready by tomorrow).

0dysseas commented:
Thanks @cdrini! I will pull the new image tonight and use the function in my PR. By the way do you know how I can get the topmost Internet Archive black bar to be displayed in my local setup or should we wait for @mekarpeles  input on this?

cdrini commented:
The design is good as is for now for this PR :) Adding a drop down is something we might want to do as a follow-up. (In a bizarre twist of fate, the IA bar has disappeared from our production site!)

cclauss commented:
Conflicts must be resolved.

0dysseas commented:
Updated my solution to use the `i18n_loadhook` infogami function instead of the `get_ol_locale`. The Open Library locale is now determined using this function.

Also, changed the `js` code on the client-side according to the review comments while keeping the UI design as-is (i.e. drop-down menu on the footer).

cdrini commented on openlibrary/core/helpers.py:
```diff
@@ -33,6 +33,7 @@
 # handy utility to parse ISO date strings
 from infogami.infobase.utils import parse_datetime
 from infogami.utils.view import safeint
+from infogami.utils.i18n import i18n_loadhook
```
We don't need to import this anywhere; this function is automagically called every time a request is made (since it's registered here: https://github.com/internetarchive/infogami/blob/8e03b4fb0d844ded794112e90d08da9787ed1ba4/infogami/utils/i18n.py#L200 )
```suggestion
```

cdrini commented on openlibrary/core/helpers.py:
```diff
@@ -48,7 +49,7 @@
 
     # functions imported from elsewhere
-    "parse_datetime", "safeint"
+    "parse_datetime", "safeint", "i18n_loadhook"
```
```suggestion
    "parse_datetime", "safeint"
```

cdrini commented on openlibrary/i18n/__init__.py:
```diff
@@ -8,6 +8,7 @@
 from babel.messages import Catalog
 from babel.messages.pofile import read_po, write_po
 from babel.messages.mofile import write_mo
+from infogami.utils.i18n import i18n_loadhook
```
```suggestion
```

cdrini commented on openlibrary/i18n/__init__.py:
```diff
@@ -111,7 +112,9 @@ def load_locale(lang):
         """Translate a given string to the language of the current locale."""
-        translations = load_translations(web.ctx.get('lang', 'en'))
+        # Get the website locale from the global ctx.lang variable, set in i18n_loadhook
+        website_locale = web.ctx.lang        
+        translations = load_translations(website_locale)
```
Extra variable reduces readability here a bit :) The comment says it all!

```suggestion
        translations = load_translations(web.ctx.lang)
```

cdrini commented on openlibrary/i18n/__init__.py:
```diff
@@ -147,9 +150,12 @@ def __add__(self, other):
-    translations = load_translations(web.ctx.get('lang', 'en'))
-    value = (translations and translations.ungettext(s1, s2, _n))
+    # Get the website locale from the global ctx.lang variable, set in i18n_loadhook
+    website_locale = web.ctx.lang  
+    translations = load_translations(website_locale)
```
```suggestion
    translations = load_translations(web.ctx.lang)
```

cdrini commented on openlibrary/i18n/__init__.py:
```diff
@@ -165,9 +171,10 @@ def ungettext(s1, s2, _n, *a, **kw):
-    locale = load_locale(web.ctx.get('lang', 'en'))
+    """Returns the territory name in the current locale."""
+    # Get the website locale from the global ctx.lang variable, set in i18n_loadhook
+    lang = web.ctx.lang  
+    locale = load_translations(lang)
     return locale.territories.get(code, code)
```
Typo; this one's calling a different function.

```suggestion
    locale = load_locale(web.ctx.lang)
    return locale.territories.get(code, code)
```

cdrini commented on openlibrary/i18n/messages.pot:
```diff
@@ -2556,6 +2556,10 @@ msgstr ""
 msgid "Suggesting Edits"
 msgstr ""
 
+#: nav_foot.html:51
+msgid "Change Website Language"
+msgstr ""
+
```
These are automatically generated later; let's revert the changes to this file; it looks like the file permissions also changed somehow?

0dysseas replied:
Done.

cdrini commented on openlibrary/plugins/openlibrary/home.py:
```diff
@@ -4,7 +4,7 @@
 import logging
 
-from infogami.utils import delegate
+from infogami.utils import delegate, i18n
```
```suggestion
from infogami.utils import delegate
```

cdrini commented on openlibrary/plugins/openlibrary/home.py:
```diff
@@ -64,7 +63,10 @@ class home(delegate.page):
 
     def GET(self):
-        cached_homepage = get_cached_homepage()
+        # Set the global web.ctx.lang var to the website locale that will be used
+        i18n.i18n_loadhook()
+        website_locale = web.ctx.lang
+        cached_homepage = get_cached_homepage(website_locale)
```
```suggestion
        cached_homepage = get_cached_homepage()
```

cdrini commented on openlibrary/plugins/openlibrary/home.py:
```diff
@@ -50,9 +49,9 @@ def get_homepage():
-def get_cached_homepage():
+
+def get_cached_homepage(lang):
     five_minutes = 5 * dateutil.MINUTE_SECS
```
```suggestion
    five_minutes = 5 * dateutil.MINUTE_SECS
    lang = web.ctx.lang
```

cdrini commented on openlibrary/plugins/openlibrary/js/ol.js:
```diff
@@ -147,3 +155,15 @@ export function initPreviewButton() {
 }
+
+export function initWebsiteTranslationOptions() {
+    $(document).ready(function () {
+        $('#locale-options li a').on('click', function () {
+            event.preventDefault();
+            const locale = $(this).data('lang-id');
+            setValueInCookie('HTTP_LANG', locale);
+            location.reload();
+        });
+    });
```
- Don't need `document.ready`; that's already happened at this point

```suggestion
    $('#locale-options li a').on('click', function (event) {
        event.preventDefault();
        const locale = $(this).data('lang-id');
        setValueInCookie('HTTP_LANG', locale);
        location.reload();
    });
```

cdrini commented on openlibrary/templates/lib/nav_foot.html:
```diff
@@ -47,6 +47,17 @@ <h2>$:_('Help')</h2>
         </aside>
       </div>
+      <div id="footer-locale-menu">
+        <h2>$:_('Change Website Language')</h2> 
```
In the web.py templating system, `$` does HTML escaping, whereas `$:` does not. So `$` is preferred as the default unless we know we're going to have some HTML! (E.g. `$:_("<a>Foo</a>")`)

```suggestion
        <h2>$_('Change Website Language')</h2> 
```

cdrini commented on openlibrary/templates/lib/nav_foot.html:
```diff
@@ -47,6 +47,17 @@ <h2>$:_('Help')</h2>
+      <div id="footer-locale-menu">
+        <h2>$:_('Change Website Language')</h2> 
+        <ul id="locale-options">
+          <li><a href="#" data-lang-id="cs" title="Čeština">Čeština (cs)</a> </li>
+          <li><a href="#" data-lang-id="de" title="Deutsch">Deutsch (de)</a> </li>
+          <li><a href="#" data-lang-id="en" title="English">English (en)</a></li>
+          <li><a href="#" data-lang-id="es" title="Español">Español (es)</a></li>
+          <li><a href="#" data-lang-id="fr" title="Français">Français (fr)</a></li>
+          <li><a href="#" data-lang-id="te" title="తెలుగు">తెలుగు (te)</a></li>
```
- Thinking about it some more, I think I like your original approach best of keeping the translated language name in the title. That's also what Wikipedia does.
- Added Croatian
```suggestion
          <li><a href="#" data-lang-id="cs" title="$_('Czech')">Čeština (cs)</a> </li>
          <li><a href="#" data-lang-id="de" title="$_('German')">Deutsch (de)</a> </li>
          <li><a href="#" data-lang-id="en" title="$_('English')">English (en)</a></li>
          <li><a href="#" data-lang-id="es" title="$_('Spanish')">Español (es)</a></li>
          <li><a href="#" data-lang-id="fr" title="$_('French')">Français (fr)</a></li>
          <li><a href="#" data-lang-id="hr" title="$_('Croatian')">Hrvatski (hr)</a></li>
          <li><a href="#" data-lang-id="te" title="$_('Telugu')">తెలుగు (te)</a></li>
```

cdrini commented on openlibrary/templates/type/doc/edit.html:
```diff
@@ -1,6 +1,5 @@
 $def with (page)
 
-$ lang = i18n.get_locale() or 'en'
```
Let's revert these 3 HTML files; these are _very old_ templates, which used a different system for i18n, I believe.

0dysseas replied:
Done.

cdrini commented on openlibrary/i18n/__init__.py:
```diff
@@ -8,6 +8,7 @@
 from babel.messages import Catalog
```
File permissions have changed on this file; let's change them back.

0dysseas replied:
Done.

cdrini reviewed (changes requested):
Awesome @0dysseas ! This is almost there :) A few final comments added.

cdrini commented:
I applied all the smaller changes view the GitHub UI; just some file permission cleanup left!

cdrini commented on openlibrary/plugins/openlibrary/home.py:
```diff
@@ -50,9 +49,10 @@ def get_homepage():
 
-def get_cached_homepage():
+
+def get_cached_homepage(lang):
```
Ooops! Meant this to be part of the previous changes.
```suggestion
def get_cached_homepage():
```

cdrini reviewed (approved):
Ok! Was noticing staging erroring on the homepage kind of randomly; so figured out what that was about! Turns out it was related to the homepage being rendered on a different thread sometimes. But now this PR also fixes an old issue of the homepage sometimes randomly showing up in English even if ?lang is set to something else! \o/

0dysseas commented:
Looking forward to seeing this in production too @cdrini! I'm glad I could help open library!

cclauss commented:
Awesome work!!!
